src/compiler/checker.ts

```diff
@@ -12701,6 +12701,19 @@ namespace ts {
                     return anyType;
                 }
                 if (accessExpression && !isConstEnumObjectType(objectType)) {
+                    if (isObjectLiteralType(objectType)) {
+                        if (indexType.flags & (TypeFlags.StringLiteral | TypeFlags.NumberLiteral)) {
+                            return undefinedType;
+                        }
+                        else if (indexType.flags & (TypeFlags.Number | TypeFlags.String)) {
+                            const types = (<ResolvedType>objectType).properties
+                                .map(property => {

```

Comment:
Uh, minor nit, use `ts.map` instead of array's builtin `map`. Mostly because of stylistic consistency at this point.